File: algorithms/03_cpp/sorting/insertion_sort.cpp
/**
 * Insertion sort.
 */

#include <vector>
#include <algorithm>
#include <cassert>
#include <iostream>

void insertionSort(std::vector<int>& a) {
    for (int i = 1; i < static_cast<int>(a.size()); ++i) {
        int key = a[i];
        int j = i - 1;
        while (j >= 0 && a[j] > key) {
            a[j + 1] = a[j];
            --j;
        }
        a[j + 1] = key;
    }
}

int main() {
    std::vector<int> data = {5, 2, 9, 1, 5, 6};
    insertionSort(data);
    assert(std::is_sorted(data.begin(), data.end()));
    assert(data.front() == 1 && data.back() == 9);
    std::cout << "[C++ InsertionSort] Verified." << std::endl;
    return 0;
}

File: algorithms/04_java/sorting/InsertionSort.java
import java.util.Arrays;

public class InsertionSort {
    public static void sort(int[] a) {
        for (int i = 1; i < a.length; i++) {
            int key = a[i];
            int j = i - 1;
            while (j >= 0 && a[j] > key) {
                a[j + 1] = a[j];
                j--;
            }
            a[j + 1] = key;
        }
    }

    public static void main(String[] args) {
        int[] data = {33, 7, 91, 12, 5, 5, 78, 2, 44, 19};
        sort(data);
        for (int i = 1; i < data.length; i++)
            if (data[i - 1] > data[i]) throw new AssertionError("not sorted at " + i);
        System.out.println("[Java InsertionSort] Insertion sort verified: " + Arrays.toString(data));
    }
}
